### PR TITLE
[ML] Create task for model download

### PR DESCRIPTION
**Pull request by jonathan-buttner**
This PR adds the functionality to create a task when a model is being downloaded. The task is cancellable via the cancellation api: https://www.elastic.co/guide/en/elasticsearch/reference/current/tasks.html#task-cancellation

This PR doesn't add any cleaning up after a task is cancelled. Future PRs will add:
- The put trained model API will allow overwriting an existing elser model instead of throwing an error that it already exists, this will allow the user to retry the download without needing to delete it first (https://github.com/elastic/elasticsearch/pull/97733)
- The delete api will check if the download is currently running and cancel the task

## Testing

To test that I could successfully cancel the download, I had to do a couple things. First I used the mac Network Link Conditioner to simulate a slow network. This increases the download time so that we can hit the cancellation api.

https://apple.stackexchange.com/questions/24066/how-to-simulate-slow-internet-connections-on-the-mac

To use that you'll need to download the `Additional Tools for Xcode`:

[Additional Tools for Xcode 14.3 Release Candidate.dmg](https://download.developer.apple.com/Developer_Tools/Additional_Tools_for_Xcode_14.3_Release_Candidate/Additional_Tools_for_Xcode_14.3_Release_Candidate.dmg)

You'll need a developer account with apple to do that but it should be free.

Kick off the download:

```
curl --location --request PUT 'http://localhost:9200/_ml/trained_models/.elser_model_1?wait_for_completion=true' \
--header 'Content-Type: application/json' \
--header 'Authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
--data '{"input":{"field_names":["text_field"]}}'
```

Cancel the download

```
curl --location --globoff --request POST 'http://localhost:9200/_tasks/_cancel?actions=*model_import[n]' \
--header 'Authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
--data ''
```

The API request should return with an error stating that it was cancelled, it may take a bit depending on how slow the simulated connection is.

### CODE REVIEW

File: x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java [hunk 1]
```diff
@@ -72,26 +79,49 @@ public TransportLoadTrainedModelPackage(
...
+    static void importModel(
+        Client client,
+        TaskManager taskManager,
+        Request request,
+        ModelImporter modelImporter,
+        ActionListener<AcknowledgedResponse> listener,
+        Task task
+    ) {
         String modelId = request.getModelId();
         final AtomicReference<Exception> exceptionRef = new AtomicReference<>();
```
Comment (hendrikmuhs) [inline x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java:126]: as `importModel` takes over ownership of the task these 2 lines should be part of `try {`

File: x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java [hunk 2]
```diff
@@ -72,26 +79,49 @@ public TransportLoadTrainedModelPackage(
...
+        ModelImporter modelImporter = new ModelImporter(
+            parentTaskAssigningClient,
+            request.getModelId(),
+            request.getModelPackageConfig(),
+            downloadTask
+        );
 
         threadPool.executor(MachineLearningPackageLoader.UTILITY_THREAD_POOL_NAME)
-            .execute(() -> importModel(client, request, modelImporter, listener));
+            .execute(() -> importModel(client, taskManager, request, modelImporter, listener, downloadTask));
```
[the remark this answers is not in the document]
Comment (hendrikmuhs) [reply x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java:101]: can you put that into a code doc? In the places where it is possible, maybe call it e.g. `auditClient`, so it is more clear that this _client_ is only for the notifications.

File: x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java [hunk 3]
```diff
@@ -124,14 +131,25 @@ private void uploadVocabulary() throws URISyntaxException {
...
             vocabularyAndMerges.v2(),
             List.of()
         );
 
-        client.execute(PutTrainedModelVocabularyAction.INSTANCE, r2).actionGet();
+        executeRequest(PutTrainedModelVocabularyAction.INSTANCE, request);
+    }
+
+    private <Request extends ActionRequest, Response extends ActionResponse> void executeRequest(
+        ActionType<Response> action,
```
Comment (hendrikmuhs) [inline x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java:145]: nit: maybe call it `executeRequestIfNotCancelled` ?

File: x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonathan-buttner) [reply x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java:101]: Good idea, will do 👍

**Review by hendrikmuhs — APPROVED** [review]
LGTM

**Pull request merged**